gpMgmt/bin/gprebalance_modules/solver.py

```diff
@@ -0,0 +1,1063 @@
+                subset = {seg_id: assignments[seg_id] for seg_id in subset_keys}
+                subset_hash = self._hash_assignment(subset)
+                
+                if subset_hash in self.nogoods:
+                    self.hits += 1
+                    return True
+        
+        return False
+    
+    def _hash_assignment(self, assignments: Dict[int, Tuple[int, int]]) -> str:
+        """Hash a partial assignment for storage"""
+        return str(sorted(assignments.items()))
+
+# ============================================================================
+# SEARCH NODE - State in the Branch-and-Bound tree
+# ============================================================================
+
+@dataclass
+class SearchNode:
+    """
+    Node in the search tree.
+    
+    Attributes:
+        domains: Valid (primary_host, mirror_host) pairs for each segment
+        assignments: Current assignment branch {seg_id: (primary_host, mirror_host)}
+        depth: Number of segments assigned
+        lower_bound: Minimum cost achievable from this node
+        host_load: host_load[host_id] = number of segments on that host
+    """
+    domains: SegmentDomain
+    assignments: Dict[int, Tuple[int, int]]  # seg_id → (primary_host, mirror_host)
+    depth: int
+    lower_bound: int = 0
+    host_load: List[int] = field(default_factory=list)  # host_load[host_id]
+    
+    def is_complete(self, n_segments: int) -> bool:
+        """Check if all segments are assigned"""
+        return len(self.assignments) == n_segments
+    
+    def assign_segment(self, seg_id: int, primary_host: int, mirror_host: int):
+        """Assign a segment and update domain"""
+        self.assignments[seg_id] = (primary_host, mirror_host)
+        self.domains.pairs[seg_id] = {(primary_host, mirror_host)}
+
+# ============================================================================
+# DYNAMIC MIRROR GROUPS - Groups computed from current assignments
+# ============================================================================
+
+class DynamicMirrorGroups:
+    """
+    Compute mirror groups dynamically based on CURRENT primary assignments.
+    Not based on initial mirroring, but on actual assigned primaries during search.
+    """
+    
+    @staticmethod
+    def get_groups(node: SearchNode, strategy: str) -> List[Dict]:
+        """
+        Build mirror groups from CURRENT assignments.
+        
+        Returns:
+            List of groups: [{'primary_host': h, 'seg_ids': [...], 'strategy': ...}, ...]
+        """
+        groups_by_primary = defaultdict(list)
+        
+        # Group ASSIGNED segments by their NEW primary_host
+        for seg_id, (primary_host, mirror_host) in node.assignments.items():
+            groups_by_primary[primary_host].append(seg_id)
+        
+        # Convert to group structures (only if 2+ segments share primary)
+        mirror_groups = []
+        for primary_host in sorted(groups_by_primary.keys()):
```

Comment:
In the test example from the description, `groups_by_primary.keys()` returns the key "default_factory" is this expected?